utils/cache/blob_memory_cache.go

```diff
@@ -135,12 +143,41 @@ func (c *BlobMemoryCache) NumEntries() int {
 }
 
 // TotalBytes returns the current total size in bytes.
-func (c *BlobMemoryCache) TotalBytes() int64 {
+func (c *BlobMemoryCache) TotalBytes() uint64 {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 	return c.totalSize
 }
 
+// TryReserve attempts to reserve space for buffering before adding to cache.
+// Returns true if space was reserved, false if insufficient space.
+func (c *BlobMemoryCache) TryReserve(size uint64) bool {
+	for {
+		total := c.TotalBytes()
+
+		reserved := c.reservedSize.Load()
+		if total+reserved+size > c.config.MaxSize {
+			return false
+		}
+		if c.reservedSize.CompareAndSwap(reserved, reserved+size) {
+			return true
+		}
+	}
+}
+
+// ReleaseReservation releases previously reserved space.
+func (c *BlobMemoryCache) ReleaseReservation(size uint64) {
+	for {
+		reserved := c.reservedSize.Load()
+		if reserved == 0 || size > reserved {
+			return // Nothing to release
```

Comment:
This is unexpected, right? Getting to this point means we have an implementation error. Let's emit an error log? Something like the above?

Also, if we decide to fail-open here, let's do `if reserved == 0 {return} else if size > reserved {reserved = 0; return}` instead of the current implementation to prevent a potential leak. WDYT?
```suggestion
            c.log.Error("invariant violation: ReleaseReservation called but reserved memory is smaller than argument")
			return // Nothing to release
```

Reply:
I agree. Let's add the log